File: src/main/java/com/team195/CKSim.java
package com.team195;

import com.team195.jni.CKSimJNI;
import java.util.HashSet;

public class CKSim {
    private static CKSimJNI m_jniInstance = new CKSimJNI();
    private static CKSim mInstance = new CKSim();

    public static final int MAX_NUM_MOTORS = 64;
    public static final int MAX_NUM_ACCEL = 64;
    public static final int MAX_NUM_GYRO = 64;
    public static final int MAX_NUM_ADVOBJ = 64;

    private final HashSet<Integer> motorReg;
    private final HashSet<Integer> accelReg;
    private final HashSet<Integer> gyroReg;
    private final HashSet<Integer> advObjReg;

    protected static CKSim getInstance() {
        return mInstance;
    }

    private CKSim() {
        motorReg = new HashSet<>();
        accelReg = new HashSet<>();
        gyroReg = new HashSet<>();
        advObjReg = new HashSet<>();
    }

    public boolean registerMotor(int id) {
        return motorReg.add(id);
    }

    public boolean containsMotor(int id) {
        return motorReg.contains(id);
    }

    public boolean registerAccelerometer(int id) {
        return accelReg.add(id);
    }

    public boolean containsAccelerometer(int id) {
        return accelReg.contains(id);
    }

    public boolean registerGyro(int id) {
        return gyroReg.add(id);
    }

    public boolean containsGyro(int id) {
        return gyroReg.contains(id);
    }

    public boolean registerAdvObj(int id) {
        return advObjReg.add(id);
    }

    public boolean containsAdvObj(int id) {
        return advObjReg.contains(id);
    }

    public float getMotorValue(int id) {
        return CKSimJNI.GetMotorValue(id);
    }

    public void setMotorValue(int id, float val) {
        CKSimJNI.SetMotorValue(id, val);
    }

    public float getEncoderValue(int id) {
        return CKSimJNI.GetEncoderValue(id);
    }

    public void setEncoderValue(int id, float val) {
        CKSimJNI.SetEncoderValue(id, val);
    }

    public float getAccelerometerValue(int id) {
        return CKSimJNI.GetAccelerometerValue(id);
    }

    public void setAccelerometerValue(int id, float val) {
        CKSimJNI.SetAccelerometerValue(id, val);
    }

    public float getGyroValue(int id) {
        return CKSimJNI.GetGyroValue(id);
    }

    public void setGyroValue(int id, float val) {
        CKSimJNI.SetGyroValue(id, val);
    }

    public float getAdvObjValue(int id) {
        return CKSimJNI.GetAdvObjValue(id);
    }

    public void setAdvObjValue(int id, float val) {
        CKSimJNI.SetAdvObjValue(id, val);
    }
}

File: src/main/native/cpp/CKSim.cpp
#include "team195/CKSim.h"

#include "team195/CKSimDriver.h"
#include <unordered_set>

namespace team195
{

    CKSim::CKSim()
    {
        if (c_CKSimDriver() < 0)
        {
            throw std::exception("CKSim Driver Init failed!");
        }
    }

    CKSim::~CKSim()
    {
        c_CKSimDealloc();
    }

    void CKSim::SetIP(std::string ip)
    {
        c_CKSimSetIP(ip.c_str());
    }

    static std::unordered_set<int> motorReg;
    bool CKSim::registerMotor(int id)
    {
        return motorReg.insert(id).second; //Return success/failure bool from pair
    }
    bool CKSim::containsMotor(int id)
    {
        return motorReg.find(id) != motorReg.end();
    }

    static std::unordered_set<int> accelReg;
    bool CKSim::registerAccelerometer(int id)
    {
        return accelReg.insert(id).second;
    }
    bool CKSim::containsAccelerometer(int id)
    {
        return accelReg.find(id) != accelReg.end();
    }

    static std::unordered_set<int> gyroReg;
    bool CKSim::registerGyro(int id)
    {
        return gyroReg.insert(id).second;
    }
    bool CKSim::containsGyro(int id)
    {
        return gyroReg.find(id) != gyroReg.end();
    }

    static std::unordered_set<int> advObjReg;
    bool CKSim::registerAdvObj(int id)
    {
        return advObjReg.insert(id).second;
    }
    bool CKSim::containsAdvObj(int id)
    {
        return advObjReg.find(id) != advObjReg.end();
    }

    float CKSim::getMotorValue(int id)
    {
        return c_GetMotor(id);
    }
    void CKSim::setMotorValue(int id, float val)
    {
        c_SetMotor(id, val);
    }

    float CKSim::getEncoderValue(int id)
    {
        return c_GetEncoder(id);
    }
    void CKSim::setEncoderValue(int id, float val)
    {
        c_SetEncoder(id, val);
    }

    float CKSim::getAccelerometerValue(int id)
    {
        return c_GetAccelerometer(id);
    }
    void CKSim::setAccelerometerValue(int id, float val)
    {
        c_SetAccelerometer(id, val);
    }

    float CKSim::getGyroValue(int id)
    {
        return c_GetGyro(id);
    }
    void CKSim::setGyroValue(int id, float val)
    {
        c_SetGyro(id, val);
    }

    float CKSim::getAdvObjValue(int id)
    {
        return c_GetAdvObj(id);
    }
    void CKSim::setAdvObjValue(int id, float val)
    {
        c_SetAdvObj(id, val);
    }
} // namespace team195